src/NHibernate/Transaction/AdoNetWithSystemTransactionFactory.cs

```diff
@@ -0,0 +1,528 @@
+			}
+
+			/// <summary>
+			/// Unlock the context, causing <see cref="Wait"/> to cease blocking. Do nothing if the context
+			/// is not locked.
+			/// </summary>
+			protected virtual void Unlock()
+			{
+				// Do not set _locked back to false, we want to wait only once per transaction context.
+				// Volatile member, do not simplify by removing the local variable.
+				var semaphore = _semaphore;
+				_semaphore = null;
+				if (semaphore == null)
+					return;
+				// Not supposed to have more than one waiting thread, but in case the user made bad multi-threading,
+				// do not leave frozen threads.
+				semaphore.Release(100);
+				semaphore.Dispose();
+			}
+
+			#region IEnlistmentNotification Members
+
+			/// <summary>
+			/// Prepare the session for the transaction commit. Runs
+			/// <see cref="ISessionImplementor.BeforeTransactionCompletion(ITransaction)"/> for the session and for
+			/// <see cref="ConnectionManager.DependentSessions"/> if any. <see cref="Lock"/> the context
+			/// before signaling it is done, or before rollback in case of failure.
+			/// </summary>
+			/// <param name="preparingEnlistment">The object for notifying the prepare phase outcome.</param>
+			public virtual void Prepare(PreparingEnlistment preparingEnlistment)
+			{
+				using (new SessionIdLoggingContext(_sessionImplementor.SessionId))
+				{
+					try
+					{
+						using (_sessionImplementor.ConnectionManager.BeginsProcessingFromSystemTransaction(_useConnectionOnSystemTransactionEvents))
+						{
+							if (_useConnectionOnSystemTransactionEvents)
+							{
+								// Ensure any newly acquired connection gets enlisted in the transaction. When distributed,
+								// this code runs from another thread and we cannot rely on Transaction.Current.
+								using (var tx = new TransactionScope(AmbientTransaction))
+								{
+									_sessionImplementor.BeforeTransactionCompletion(null);
+									foreach (var dependentSession in _sessionImplementor.ConnectionManager.DependentSessions)
+										dependentSession.BeforeTransactionCompletion(null);
+
+									tx.Complete();
+								}
+							}
+							else
+							{
+								_sessionImplementor.BeforeTransactionCompletion(null);
+								foreach (var dependentSession in _sessionImplementor.ConnectionManager.DependentSessions)
+									dependentSession.BeforeTransactionCompletion(null);
+							}
+						}
+						// Lock the session to ensure second phase gets done before the session is used by code following
+						// the transaction scope disposal.
+						Lock();
+
+						_logger.Debug("Prepared and done for system transaction");
+						// We do not have anything more to do in second phases callbacks: the remaining work is handled in
+						// transaction completion event, which always executes, contrary to second phases callbacks which
+						// do not, depending on the rollback cases.
+						// This saves a thread when distributed.
+						preparingEnlistment.Done();
```

Comment:
The NHibernate volatile do all its work on its prepare phase and has nothing to do on the second phase. For performances, it is then advised to call `Done` instead of `Prepared`. It tells the transaction coordinator it does not have to call second phase on that resource. When distributed, this saves a thread by the way.